scrapers/utils.py

```diff
@@ -134,3 +137,77 @@ def validate_and_get_download_link(base_url, download_link, check_content_pdf=Tr
             raise TypeError("File extension does not match content type.")
 
     return absolute_url
+
+
+def calculate_checksum(filepaths):
+    md5 = hashlib.md5()
+    for file in filepaths:
+        try:
+            md5.update(Path(file).read_bytes())
+        except IsADirectoryError:
+            pass
+    return md5.hexdigest()
+
+
+def zip_pdfs(scraper_filepath, filepaths, link_id):
+    with download_directory(scraper_filepath) as directory:
+        zip_name = get_standardized_filename(scraper_filepath, link_id, ".zip")
+        zip_path = get_file_path(directory, zip_name)
+
+    try:
+        with zipfile.ZipFile(zip_path, "w") as zipf:
+            for file in filepaths:
+                filename = os.path.basename(file)
+                zipf.write(file, arcname=filename)
+            for file in filepaths:
+                try:
+                    if os.path.exists(file):
+                        os.remove(file)
+                    else:
+                        print(f"File '{file}' does not exist.")
+                except Exception as e:
+                    print(f"Error deleting file '{file}': {e}")
+    except Exception as e:
+        print(f"Error creating zip file: {e}")
+
+    return zip_name, zip_path
+
+
+def connect_circulars_to_associated_files(file_names):
+    """
+    Accepts 'file_names' as an argument, which is a list of lists in which each sub-list consists of a file name at index 0 and a file extension at index 1.
+
+    This function identifies files that are part of the same circular and groups them together in a returned dictionary, in which each key is a circular name and its value is a list of all files belonging to that circular.
+    """
+    circular_to_links_dict = {}
+
+    # sort file_names list by file name string length
+    file_names.sort(key=lambda x: len(x[0]))
+
+    for i, [curr_file, extension] in enumerate(file_names):
+        # by default, we add the current file as a key to the dictionary
+        key_to_be_added = curr_file
+
+        # iterate through the files added thus far to the dictionary
+        for circular in file_names[:i]:
+            # if the current file name is an extension of a file already added as a key to the dictionary, reassign the key to be added and break the loop
+            if circular[0] in curr_file:
+                key_to_be_added = circular[0]
+                break
+
+        # add the current file to the list of the 'key_to_be_added' key in the dictionary
+        circular_to_links_dict.setdefault(key_to_be_added, []).append(f"{curr_file}.{extension}")
```

Comment:
Trying to loop through part of the iterator that we were already trying to loop through in the outer loop is not the best practice. We already sort the file names in L185, correct? That means we can look through the keys of the dictionary to find the key that we want to add to, correct?

```suggestion
    for i, [curr_file, extension] in enumerate(file_names):
        keys = [key for key in circular_to_links_dict if re.search(key, curr_file)]
        if (key_length := len(keys)) > 1:
            raise ValueError(f"Expected 1 matched key for the file, found {key_length}")
        key_to_be_added = keys[0] if keys else curr_file

        # add the current file to the list of the 'key_to_be_added' key in the dictionary
        circular_to_links_dict.setdefault(key_to_be_added, []).append(f"{curr_file}.{extension}")
```

Reply:
This makes sense that we shouldn't use the same iterator twice in a nested loop. I like this change, but I wonder if it makes sense to retrieve a list of keys rather than just 1 key because the current logic makes it impossible for there to be more than 1 matching key in the dictionary.